src/bytes.rs

```diff
@@ -1,26 +1,126 @@
+    #[inline]
+    fn eq(&self, other: &Self) -> bool {
+        let this = self.as_bytes();
+        let that = other.as_bytes();
+        this == that
+    }
+}
+
+impl<'a> Hash for Bytes<'a> {
+    #[inline]
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        // Hash must be implemented manually for Bytes, otherwise it would only hash a pointer
+        let this = self.as_bytes();
+        this.hash(state);
+    }
+}
+
+impl<'a> Clone for Bytes<'a> {
+    fn clone(&self) -> Self {
+        // It is important to manually implement Clone for Bytes,
+        // because if `self` was owned, then the default clone
+        // implementation would only clone the pointer
+        // which leads to aliasing boxes, and later, when `Bytes` is dropped,
+        // the box is freed twice!
+        match &self.data {
+            BytesInner::Borrowed(data, len) => {
+                Bytes::from(unsafe { compact_bytes_to_slice(*data, *len) })
+            }
+            BytesInner::Owned(data, len) => {
+                let (ptr, len) = unsafe { clone_compact_bytes_parts(*data, *len) };
+                Bytes {
+                    data: BytesInner::Owned(ptr, len),
+                    _lt: PhantomData,
+                }
+            }
+        }
+    }
+}
 
 impl<'a> From<&'a str> for Bytes<'a> {
     #[inline]
     fn from(s: &'a str) -> Self {
-        Self(s.as_bytes())
+        <Self as From<&'a [u8]>>::from(s.as_bytes())
     }
 }
 
 impl<'a> From<&'a [u8]> for Bytes<'a> {
     #[inline]
     fn from(s: &'a [u8]) -> Self {
-        Self(s)
+        Bytes {
+            data: BytesInner::Borrowed(s.as_ptr(), s.len() as u32),
+            _lt: PhantomData,
+        }
     }
 }
 
-/// Custom `Debug` trait is implemented which displays the data as a UTF8 string,
-/// to make it easier to read for humans when logging
+/// Converts `Bytes` raw parts to a slice
+#[inline]
+unsafe fn compact_bytes_to_slice<'a>(ptr: *const u8, l: u32) -> &'a [u8] {
+    std::slice::from_raw_parts(ptr, l as usize)
+}
+
+/// Converts `Bytes` raw parts to a boxed slice
+#[inline]
+unsafe fn compact_bytes_to_boxed_slice(ptr: *mut u8, len: u32) -> Box<[u8]> {
+    let len = len as usize;
+
+    // carefully reconstruct a `Box<[u8]>` from the raw pointer and length
+    Vec::from_raw_parts(ptr, len, len).into_boxed_slice()
+}
+
+/// Converts a boxed byte slice to compact raw parts
+///
+/// The caller is responsible for freeing the returned pointer
+unsafe fn boxed_slice_to_compact_parts(slice: Box<[u8]>) -> (*mut u8, u32) {
+    // wrap box in `ManuallyDrop` so it's not dropped at the end of the scope
+    let mut slice = ManuallyDrop::new(slice);
+    let len = slice.len();
+    let ptr = slice.as_mut_ptr();
+
+    (ptr, len as u32)
+}
+
+/// Clones compact byte parts and returns the new parts
+#[inline]
+unsafe fn clone_compact_bytes_parts(ptr: *mut u8, len: u32) -> (*mut u8, u32) {
+    boxed_slice_to_compact_parts(compact_bytes_to_boxed_slice(ptr, len).clone())
```

Comment:
note to self: this calls Drop code of the reconstructed `Box<[u8]>`, which means that calling this function invalidates `ptr` (which is not okay). We must not run the destructor of the returned box by `compact_bytes_to_boxed_slice` here. It is probably better to do `compact_bytes_to_slice(...).clone()` here and avoid reconstructing a box at all.